### ISSUE TITLE
Interface iheritance: use abstract types rather than symbols for optional components?

### ISSUE THREAD

**Issue by rafaqz** [rafaqz/Interfaces.jl#33]
Each optional component would (hidden in the macro) define the abstract type like:

```julia
abstract type SomeInterface_option <: SomeInterface end
```

We could make it nicer looking, but that's the direct translation from what we have now

First we can just inherit the main interface

```julia
@interface NewInterface <: BaseInterface NewObject conditions 
```

But we could also inherit the optional component, which itself inherits the interface:

```julia
@interface NewInterface <: BaseInterface_opional_thing NewObject conditions 
```

We would test the interface by iterating over `supertype` until we hit `AbstractInterface` then working backwards, so the basal interface is tested first. When `@@implements` has multiple options, we would take the `union` of the interfaces to test at each step to avoid duplicating any tests..

**_But..._** what we really want is a hack for multiple inheritance:

```julia
@interface NewInterface <: Union{BaseInterface_option,SomeInterface} NewType conditions
```

Where we would drop the real inheritance and instead write something like:

```julia
abstract type NewInterface{T} <: Interface{T} end
inherits(::Type{<:NewInterface}) = (BaseInterface_option, SomeInterface)
```

Then all tests and trait queries would recurse over `inherits(T)` if a trait is false for `T`.

**Comment by gdalle** [comment]
I agree that inheriting interfaces is a key ingredient (see our recent [Discourse discussion](https://discourse.julialang.org/t/did-julia-community-do-something-to-improve-its-correctness/102515/209)).
Note that it is already possible at the moment, simply by adding `Interfaces.test` of the parent interface as a mandatory condition. So the key question is what we hope to achieve with a different syntax / implementation.

**Comment by rafaqz** [comment]
Yes its possible in the testing, but you don't get any traits to check that you passed it. The idea is that both tests and traits would be inherited automatically.

**Comment by rafaqz** [comment]
Sure, I guessed asuch.

But: how do we get the `NamedTuple` keys into the macro as an expression. We cant inspect a variable in a macro and we cant define types later at runtime.

**Comment by gdalle** [comment]
If there's a choice I definitely go for inheritance.
